app/queries/drug_stocks_query.rb

```diff
@@ -35,152 +32,123 @@ def drug_consumption_report
     Rails.cache.fetch(drug_consumption_cache_key,
       expires_in: ENV.fetch("ANALYTICS_DASHBOARD_CACHE_TTL"),
       force: RequestStore.store[:bust_cache]) do
-      {all: drug_consumption_totals,
-       facilities: drug_consumption_report_for_facilities,
+      {total_patients: total_patients,
+       all_drug_consumption: all_drug_consumption,
+       drug_consumption_by_facility_id: drug_consumption_by_facility_id,
+       patient_counts_by_facility_id: patient_counts_by_facility_id,
        last_updated_at: Time.now}
     end
   end
 
-  def patient_counts
-    @patient_counts ||= Patient.where(assigned_facility_id: @facilities).group(:assigned_facility).count
+  memoize def drugs
+    @protocol.protocol_drugs.where(stock_tracked: true).load
   end
 
-  def drug_stocks
-    @drug_stocks ||= DrugStock.latest_for_facilities(@facilities, @for_end_of_month).group_by(&:facility_id)
+  memoize def protocol_drugs_by_category
+    drugs
+      .sort_by(&:sort_key)
+      .group_by(&:drug_category)
+      .sort_by { |(drug_category, _)| drug_category }
+      .to_h
   end
 
-  def previous_month_drug_stocks
-    @previous_month_drug_stocks ||= DrugStock.latest_for_facilities(@facilities, end_of_previous_month).group_by(&:facility_id)
+  memoize def drug_categories
+    drugs.pluck(:drug_category).uniq
   end
 
-  def drug_stock_report_for_facilities
-    @stock_report_for_facilities ||= @facilities.each_with_object({}) { |facility, report|
-      report[facility.id] = drug_stock_for_facility(facility, drug_stocks[facility.id])
-    }
+  memoize def patient_counts_by_facility_id
+    Patient.where(assigned_facility_id: @facilities).group(:assigned_facility_id).count
```

Comment:
I'd be very curious to see how the code would shake out if we were able to re-use the same codepaths to craft `total_patients` as well as `patient_counts_by_facility_id`. The idea of `Patient.where(assigned_facility_id: something)` and `.count` are common to both paths. The only change is the `group` clause.

It may not be worth it to extract that just yet, given we only have two cases right now (total and by-facility). So feel free to abandon this idea if it gets gnarly.

Reply:
Same thought for the other pairs of methods in here.

Reply:
👍🏼  Will explore this when we add the block level totals.